compiler/compile/OMRCompilation.cpp

```diff
@@ -294,9 +295,17 @@ OMR::Compilation::Compilation(
    _gpuPtxCount(0),
    _bitVectorPool(self()),
    _typeLayoutMap((LayoutComparator()), LayoutAllocator(self()->region())),
-   _target(TR::Compiler->target),
    _tlsManager(*self())
    {
+   if (target != NULL)
+      {
+      _target = *target;
+      }
+   else
+      {
+      _target = TR::Compiler->target;
```

Comment:
If this is the pattern to be adopted does the visibility of TR::Compiler->target need to change to stop people accidentally using it rather than the target from the compilation? Having the two and having them potentially vary seems dangerous and likely to produce subtle bugs...

Reply:
I was suggesting to not attach a default value to `_target` and force user to specify one when they use `TR::Compilation()`.

Reply:
I think @andrewcraik's concern is around other areas of the compiler codebase using `TR::Compiler->` such as the optimizer or codegen and expecting it to be the same as `comp()->target()->`. Does that sound correct? 

Whether we specify a `_target` or not in `TR::Compilation()`, his concern will still be valid.

Reply:
@dchopra001 yup - any time there are two ways to answer seemingly the same question and they don't give the same answer you will get bugs. So I would prefer we find a way to avoid this duality or prevent the bugs before they happen through the structure of the code if possible.

Reply:
I think this was a problem before as well though. We would change the `comp()->target`'s `cpu` inside the `TR::Compilation` constructor. So if anyone was using `TR::Compiler->target.cpu` to grab information, it would end up being incorrect. I don't think the code in this PR specifically is introducing this problem.

Reply:
Oh okay, that seems like a tough issue to solve and we've already run into countless bugs related to `comp()->target()->` vs. `TR::Compiler->target`. In places where we are not doing compilation, TR::Compiler->target is still needed.

Reply:
> If this is the pattern to be adopted does the visibility of TR::Compiler->target need to change to stop people accidentally using it rather than the target from the compilation? Having the two and having them potentially vary seems dangerous and likely to produce subtle bugs...

Most of the queries to the target is done via the comp object; there are some outstanding locations where that hasn't happened (see https://github.com/eclipse/omr/issues/4518). I think my intention originally was to make the `_target` private to `TR::Compiler` but I couldn't do it because of those outstanding locations.

Reply:
I believe changes were made as part of the processor detection work to avoid using `TR::Compiler->target.cpu.getSupportsArch` and only use `comp()->target().cpu.isAtLeast` when checking for a particular architecture level support. I believe this is the only area that would be different at the moment.

But I agree, this is an issue, and I did have to deal with the above mentioned problem on Z.